File: ios/Classes/GeolocatorPlugin.m
#import "GeolocatorPlugin.h"


NSString *const METHOD_CHANNEL_NAME = @"flutter.baseflow.com/geolocator/methods";
NSString *const EVENT_CHANNEL_NAME = @"flutter.baseflow.com/geolocator/events";

@implementation GeolocatorPlugin {
    StreamLocationUpdatesTask *_streamLocationService;
    NSMutableDictionary <NSUUID *, id <TaskProtocol>> *_tasks;
}

- (void)dealloc {
    NSLog(@"[GeolocatorPlugin %s]", sel_getName(_cmd));
}

+ (void)registerWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *methodChannel = [FlutterMethodChannel methodChannelWithName:METHOD_CHANNEL_NAME binaryMessenger:registrar.messenger];
    FlutterEventChannel *eventChannel = [FlutterEventChannel eventChannelWithName:EVENT_CHANNEL_NAME binaryMessenger:registrar.messenger];
    
    GeolocatorPlugin *plugin = [[GeolocatorPlugin alloc] init];
    
    [registrar addMethodCallDelegate:plugin channel:methodChannel];
    [eventChannel setStreamHandler:plugin];
}


- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    CompletionHandler completionAction = ^(NSUUID *taskID) {
        [self->_tasks removeObjectForKey:taskID];
    };
    
    if ([call.method isEqualToString:@"getCurrentPosition"]) {
        [self executeTask:[[CurrentLocationTask alloc] initWithContext:[self buildTaskContext:call.arguments resultHandler:result] completionHandler:completionAction]];
    } else if ([call.method isEqualToString:@"getLastKnownPosition"]) {
        [self executeTask:[[LastKnownLocationTask alloc] initWithContext:[self buildTaskContext:call.arguments resultHandler:result] completionHandler:completionAction]];
    } else if ([call.method isEqualToString:@"placemarkFromAddress"]) {
        [self executeTask:[[ForwardGeocodeTask alloc] initWithContext:[self buildTaskContext:call.arguments resultHandler:result] completionHandler:completionAction]];
    } else if ([call.method isEqualToString:@"placemarkFromCoordinates"]) {
        [self executeTask:[[ReverseGeocodeTask alloc] initWithContext:[self buildTaskContext:call.arguments resultHandler:result] completionHandler:completionAction]];
    } else if ([call.method isEqualToString:@"distanceBetween"]) {
        [self executeTask:[[CalculateDistanceTask alloc] initWithContext:[self buildTaskContext:call.arguments resultHandler:result] completionHandler:completionAction]];
    } else {
        result(FlutterMethodNotImplemented);
    }
}


- (void)executeTask:(id <TaskProtocol>)task {
    if (_tasks == nil) {
        _tasks = [[NSMutableDictionary <NSUUID *, id <TaskProtocol>> alloc] init];
    }
    
    _tasks[task.taskID] = task;
    [task startTask];
}

- (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments eventSink:(FlutterEventSink)events {
    if (_streamLocationService != nil) {
        return [FlutterError errorWithCode:@"ALLREADY_LISTENING" message:@"You are already listening for location changes. Create a new instance or stop listening to the current stream." details:nil];
    }
    
    TaskContext *context = [self buildTaskContext:arguments resultHandler:events];
    
    _streamLocationService = [[StreamLocationUpdatesTask alloc] initWithContext:context completionHandler:nil];
    [_streamLocationService startTask];
    
    return nil;
}

- (FlutterError *_Nullable)onCancelWithArguments:(id _Nullable)arguments {
    if (_streamLocationService != nil) {
        [_streamLocationService stopTask];
        _streamLocationService = nil;
    }
    
    return nil;
}

- (TaskContext *)buildTaskContext:(id)arguments resultHandler:(ResultHandler)resultHandler {
    return [[TaskContext alloc] initWithArguments:arguments resultHandler:resultHandler];
}


@end

File: android/src/main/java/com/baseflow/flutter/plugin/geolocator/GeolocatorPlugin.java
package com.baseflow.flutter.plugin.geolocator;

import com.baseflow.flutter.plugin.geolocator.tasks.CalculateDistanceTask;
import com.baseflow.flutter.plugin.geolocator.tasks.ForwardGeocodingTask;
import com.baseflow.flutter.plugin.geolocator.tasks.LastKnownLocationTask;
import com.baseflow.flutter.plugin.geolocator.tasks.CurrentLocationTask;
import com.baseflow.flutter.plugin.geolocator.tasks.ReverseGeocodingTask;
import com.baseflow.flutter.plugin.geolocator.tasks.StreamLocationUpdatesTask;
import com.baseflow.flutter.plugin.geolocator.tasks.Task;
import com.baseflow.flutter.plugin.geolocator.tasks.TaskContext;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * GeolocatorPlugin
 */
public class GeolocatorPlugin implements MethodCallHandler, EventChannel.StreamHandler, OnCompletionListener {

    private static final String METHOD_CHANNEL_NAME = "flutter.baseflow.com/geolocator/methods";
    private static final String EVENT_CHANNEL_NAME = "flutter.baseflow.com/geolocator/events";

    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection") // mTasks is used to track active tasks, when tasks completes it is removed from the map
    private final Map<UUID, Task> mTasks = new HashMap<>();
    private final Registrar mRegistrar;
    private Task mStreamLocationTask;

    private GeolocatorPlugin(PluginRegistry.Registrar registrar) {
        this.mRegistrar = registrar;
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        GeolocatorPlugin geolocatorPlugin = new GeolocatorPlugin(registrar);

        final MethodChannel methodChannel = new MethodChannel(registrar.messenger(), METHOD_CHANNEL_NAME);
        final EventChannel eventChannel = new EventChannel(registrar.messenger(), EVENT_CHANNEL_NAME);
        methodChannel.setMethodCallHandler(geolocatorPlugin);
        eventChannel.setStreamHandler(geolocatorPlugin);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        TaskContext context = TaskContext.buildFromMethodResult(
                mRegistrar,
                result,
                call.arguments,
                this);

        switch (call.method) {
            case "getLastKnownPosition": {
                Task task = new LastKnownLocationTask(context);
                mTasks.put(task.getTaskID(), task);
                task.startTask();
                break;
            }
            case "getCurrentPosition": {
                Task task = new CurrentLocationTask(context);
                mTasks.put(task.getTaskID(), task);
                task.startTask();
                break;
            }
            case "placemarkFromAddress": {
                Task task = new ForwardGeocodingTask(context);
                mTasks.put(task.getTaskID(), task);
                task.startTask();
                break;
            }
            case "placemarkFromCoordinates": {
                Task task = new ReverseGeocodingTask(context);
                mTasks.put(task.getTaskID(), task);
                task.startTask();
                break;
            }
            case "distanceBetween": {
                Task task = new CalculateDistanceTask(context);
                mTasks.put(task.getTaskID(), task);
                task.startTask();
                break;
            }
            default:
                result.notImplemented();
                break;
        }
    }

    @Override
    public void onListen(Object o, EventChannel.EventSink eventSink) {
        if (mStreamLocationTask != null) {
            eventSink.error(
                    "ALREADY_LISTENING",
                    "You are already listening for location changes. Create a new instance or stop listening to the current stream.",
                    null);

            return;
        }

        TaskContext context = TaskContext.buildFromEventSink(
                mRegistrar,
                eventSink,
                o,
                this);

        mStreamLocationTask = new StreamLocationUpdatesTask(
                context);
        mStreamLocationTask.startTask();
    }

    @Override
    public void onCancel(Object arguments) {
        if (mStreamLocationTask == null) return;

        mStreamLocationTask.stopTask();
        mStreamLocationTask = null;
    }

    public void onCompletion(UUID taskID) {
        mTasks.remove(taskID);
    }


}
